src/libxtp/threecenter_gwbse.cc

```diff
@@ -150,7 +188,25 @@ void TCMatrix_gwbse::FillBlock(std::vector<Eigen::MatrixXd>& block,
       }
     }  // gamma-loop
   }    // alpha-loop
-  for (int k = 0; k < auxshell.getNumFunc(); ++k) {
+
+  return symmstorage;
+}
+
+/*
+ *  Convolution of the GW shell with ALL functions with the DFT orbital
+ * coefficients
+ */
+void TCMatrix_gwbse::FillBlock(std::vector<Eigen::MatrixXd>& block,
+                               const std::vector<Eigen::MatrixXd>& symmstorage,
+                               const Eigen::MatrixXd& dft_orbitals) {
+
+  const Eigen::MatrixXd dftm =
+      dft_orbitals.block(0, _mmin, dft_orbitals.rows(), _mtotal);
+  const Eigen::MatrixXd dftn =
+      dft_orbitals.block(0, _nmin, dft_orbitals.rows(), _ntotal);
+
+  int dim = static_cast<int>(symmstorage.size());
+  for (auto k = 0; k < dim; ++k) {
     const Eigen::MatrixXd& matrix = symmstorage[k];
     Eigen::MatrixXd threec_inMo =
         dftn.transpose() * matrix.selfadjointView<Eigen::Lower>() * dftm;
```

Comment:
How does CUDA handle the fact that of matrix only the lower triangle is filled <Eigen::Lower>?

Reply:
CUDA does not whether the structure of the matrix. the `CudaMatrix` class allocates spaces for the whole matrix and  the `copy_to_gpu` method transfer the raw data `matrix.data()` to the GPU.

Reply:
Okay, it is weird that we get the right result, because half the Eigen::MatrixXd is empty, i.e. the upper triangular part.

Reply:
Weird that we do not see that in the results

Reply:
Ahh `matrix.selfadjointView<Eigen::Lower>()` makes a copy with all entries filled in